src/db/mod.rs

```diff
@@ -19,23 +21,74 @@ pub trait Store: Clone + Sync + Send + Sized {
-    /// TODO: Create more generic error responses
-    async fn check_username_exists(&self, username: &str) -> Result<bool, Box<dyn Error>>;
+    /// checks if username exists
+    async fn check_username_exists(&self, username: &str) -> Result<bool, Error>;
 
     async fn fetch_user_id<'a>(
         &self,
         user_id: &'a UserIdentifier,
-    ) -> Result<Option<Cow<'a, UserId>>, Box<dyn Error>>;
+    ) -> Result<Option<Cow<'a, UserId>>, Error>;
 
-    async fn fetch_password_hash(&self, user_id: &UserId) -> Result<PWHash, Box<dyn Error>>;
+    async fn fetch_password_hash(&self, user_id: &UserId) -> Result<PWHash, Error>;
 
-    async fn check_otp_exists(&self, user_id: &UserId, otp: &str) -> Result<bool, Box<dyn Error>>;
+    async fn check_otp_exists(&self, user_id: &UserId, otp: &str) -> Result<bool, Error>;
 
     async fn set_device<'a>(
         &self,
         user_id: &UserId,
         device_id: &DeviceId,
         display_name: Option<&str>,
-    ) -> Result<(), Box<dyn Error>>;
+    ) -> Result<(), Error>;
+}
+
+/// Store Error
+///
+/// Errors returned from implementaitons of the `Store`
+/// trait.
+#[derive(Debug, Clone)]
+pub struct Error {
+    // TODO: Implement source error
+    pub code: ErrorCode,
+}
+
+impl std::error::Error for Error {}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.code {
+            // TODO: implment source error and chain dispaly
+            ErrorCode::ConnectionFailed => write!(f, "Connection failed."),
+            ErrorCode::AuthFailed => write!(f, "Authentication failed."),
+            ErrorCode::RecordNotFound => write!(f, "The data store could not find any records."),
+            ErrorCode::DuplicateViolation => {
+                write!(f, "A Key or Unique constraint has been violated.")
+            }
+            ErrorCode::NullViolation => write!(f, "A non-Null constraint was violated."),
+            ErrorCode::InvalidSyntax => write!(f, "The query contained invalid syntax."),
+            _ => write!(f, "An unknown error has occurred."),
+        }
+    }
+}
+
+/// A generic list of error codes returned from storge/db servers.
+///
+/// Used to simplify error handling responsed from different storage
+/// servers.  The actual implementation of the `Store` trait should use
+/// additional logging to monitor for more granular and specific errors.
+#[derive(Debug, Clone)]
+pub enum ErrorCode {
+    /// Cant connect to the storage/db server.
+    ConnectionFailed,
+    /// Authorization credentials failed.
+    AuthFailed,
+    /// Row/Record not found.
+    RecordNotFound,
+    /// A Unique Key or Duplicate Constaint check was violated.
+    DuplicateViolation,
+    /// A non-null check was violated.
+    NullViolation,
+    /// The query syntax was invalid.
+    InvalidSyntax,
+    /// Catch all for any error that can be translated to one of the existing errors.
+    Unknown,
```

Comment:
Unknown should probably contain a string, or some way of further clarifying an error. If this error should never be expected, so as much information as you can give as possible without adding bloat to non-error cases would be great

Reply:
Agreed, I think we need a better mechanism for handling `source` errors.  But haven't thought much of this.  I will tweak this.